### PR TITLE
HIVE-22256

### PR DESCRIPTION
**Pull request by jcamachor**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 1]
```diff
@@ -2463,6 +2473,15 @@ private RelNode applyPostJoinOrderingTransform(RelNode basePlan, RelMetadataProv
                 HiveProjectSortExchangeTransposeRule.INSTANCE, HiveProjectMergeRule.INSTANCE);
       }
 
+      // 10. We need to expand IN/BETWEEN expressions when loading a materialized view
```
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:2476]: Why do we need to call these rules again in post ordering?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        if (newOperands.size() > 1) {
+          return rexBuilder.makeCall(SqlStdOperatorTable.OR, newOperands);
+        }
+        return newOperands.get(0);
+      }
+      case BETWEEN: {
+        List<RexNode> newOperands = RexNodeConverter.rewriteBetweenChildren(
+            call.getOperands(), rexBuilder);
+        modified = true;
+        if (call.getOperands().get(0).isAlwaysTrue()) {
```
Comment (vineetgarg02) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java:182]: I am not sure if I understand this. In what case BETWEEN is rewritten into OR?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 1 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:2476]: This is call only when we are loading the materialized view in the registry so we can keep the plan in the correct form (IN and BETWEEN expanded) and we do not have to do it at compilation time.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java (hunk 2 above)
Comment (jcamachor) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java:182]: The first parameter to BETWEEN is a boolean flag, to differentiate BETWEEN and NOT BETWEEN.
`a BETWEEN b AND c` => `b <= a AND a <= c`
`a NOT BETWEEN b AND c` => `b > a OR a > c`

**Pull request closed without merging**